gix-blame/src/lib.rs

```diff
@@ -2,9 +2,960 @@
+/// [2]: https://github.com/git/git/commit/6dbf0c7bebd1c71c44d786ebac0f2b3f226a0131
+fn coalesce_blame_entries(lines_blamed: Vec<BlameEntry>) -> Vec<BlameEntry> {
+    // TODO
+    // It’s possible this could better be done on insertion into `lines_blamed`.
+    lines_blamed.into_iter().fold(vec![], |mut acc, entry| {
+        let previous_entry = acc.last();
+
+        if let Some(previous_entry) = previous_entry {
+            if previous_entry.commit_id == entry.commit_id
+                && previous_entry.range_in_blamed_file.end == entry.range_in_blamed_file.start
+                // As of 2024-09-19, the check below only is in `git`, but not in `libgit2`.
+                && previous_entry.range_in_original_file.end == entry.range_in_original_file.start
+            {
+                let coalesced_entry = BlameEntry {
+                    range_in_blamed_file: previous_entry.range_in_blamed_file.start..entry.range_in_blamed_file.end,
+                    range_in_original_file: previous_entry.range_in_original_file.start
+                        ..entry.range_in_original_file.end,
+                    commit_id: previous_entry.commit_id,
+                };
+
+                acc.pop();
+                acc.push(coalesced_entry);
+            } else {
+                acc.push(entry);
+            }
+
+            acc
+        } else {
+            acc.push(entry);
+
+            acc
+        }
+    })
+}
+
+// TODO: do not instantiate anything, get everything passed as argument.
+pub fn blame_file<E>(
+    odb: impl gix_object::Find + gix_object::FindHeader,
+    traverse: impl IntoIterator<Item = Result<gix_traverse::commit::Info, E>>,
+    resource_cache: &mut gix_diff::blob::Platform,
+    suspect: ObjectId,
+    worktree_path: PathBuf,
+    file_path: &BStr,
+) -> Result<Vec<BlameEntry>, E> {
+    // TODO
+    // At a high level, what we want to do is the following:
+    //
+    // - get the commit that belongs to a commit id
+    // - walk through parents
+    //   - for each parent, do a diff and mark lines that don’t have a suspect (this is the term
+    //     used in `libgit2`) yet, but that have been changed in this commit
+    //
+    // The algorithm in `libgit2` works by going through parents and keeping a linked list of blame
+    // suspects. It can be visualized as follows:
+    //
+    // <---------------------------------------->
+    // <---------------><----------------------->
+    // <---><----------><----------------------->
+    // <---><----------><-------><-----><------->
+    // <---><---><-----><-------><-----><------->
+    // <---><---><-----><-------><-----><-><-><->
+
+    // Needed for `to_str`.
+    use gix_object::bstr::ByteSlice;
+
+    let absolute_path = worktree_path.join(file_path.to_str().unwrap());
+
+    // TODO Verify that `imara-diff` tokenizes lines the same way `lines` does.
+    let number_of_lines = std::fs::read_to_string(absolute_path).unwrap().lines().count();
```

Comment:
@Byron What is the easiest way to get a file’s contents, given the commit id and the file path? In order to be able to more granularly debug `gix blame`, I’d like to add an argument that specifies the commit to start the blame at.

`git blame` has this as `<rev>` in its documentation, enabling use cases like this one: `git blame <commit_id> <file_path>` (possibly separated by `--` in case of ambiguity as far as I can tell).

Reply:
I found [Tree::lookup_entry][tree-lookup-entry] (which used to be called `lookup_path`, according to `CHANGELOG.md`), but since this is only available in `gix`, not in any of the plumbing crates, I figured it was not supposed to be used here. Which is why I’m looking for a lower-level alternative if there is one. :-)

[tree-lookup-entry]: https://docs.rs/gix/latest/gix/struct.Tree.html#method.lookup_entry

Reply:
Absolutely, in my mind I was in `gix blame` even though this is `gix-blame` the crate 🤦‍♂️. I am confused sometimes.